KAFKA-13549: Add repartition.purge.interval.ms

nicktelford opened the pull request:
See [KIP-811](https://cwiki.apache.org/confluence/display/KAFKA/KIP-811%3A+Add+config+repartition.purge.interval.ms+to+Kafka+Streams)

Repartition records are explicitly deleted once they have been fully consumed.
Currently, this is done every time the Task is committed, resulting in
"delete records" requests being sent every `commit.interval.ms`
milliseconds.

When `commit.interval.ms` is set very low, for example when
`processing.guarantee` is set to `exactly_once_v2`, this causes delete
records requests to be sent extremely frequently, potentially reducing
throughput and causing a high volume of log messages to be logged by the
brokers.

Disconnecting delete records requests from the commit interval resolves
this problem.

We now only explicitly delete records for a repartition topic when we
commit, if it's been at least `repartition.purge.interval.ms` milliseconds
since the last time we deleted records.

Because we still require a commit to trigger record deletion, the
lower-bound of `repartition.purge.interval.ms` is effectively capped at
the `commit.interval.ms`.

For compatibility, the default `repartition.purge.interval.ms` is set to
30 seconds, the same as the default `commit.interval.ms`. Users who have
configured a different `commit.interval.ms` may need to review and
change `repartition.purge.interval.ms`.

Unlike `commit.interval.ms`, we don't dynamically change the default for
`repartition.purge.interval.ms` when EOS processing is enabled, as it's
important not to flood brokers with the record deletions, and we want a
sensible default.

This code is my own work and is licensed to the Apache Kafka project under the terms of the same license (ASL 2) as the project itself.

### Committer Checklist (excluded from commit message)
- [X] Verify design and implementation 
- [X] Verify test coverage and CI build status
- [X] Verify documentation (including upgrade notes)

nicktelford committed:
KAFKA-13549: Add repartition.purge.interval.ms

See [KIP-811](https://cwiki.apache.org/confluence/display/KAFKA/KIP-811%3A+Add+config+repartition.purge.interval.ms+to+Kafka+Streams)

Repartition records are explicitly deleted once they have been fully consumed.
Currently, this is done every time the Task is committed, resulting in
"delete records" requests being sent every `commit.interval.ms`
milliseconds.

When `commit.interval.ms` is set very low, for example when
`processing.guarantee` is set to `exactly_once_v2`, this causes delete
records requests to be sent extremely frequently, potentially reducing
throughput and causing a high volume of log messages to be logged by the
brokers.

Disconnecting delete records requests from the commit interval resolves
this problem.

We now only explicitly delete records for a repartition topic when we
commit, if it's been at least `repartition.purge.interval.ms` milliseconds
since the last time we deleted records.

Because we still require a commit to trigger record deletion, the
lower-bound of `repartition.purge.interval.ms` is effectively capped at
the `commit.interval.ms`.

For compatibility, the default `repartition.purge.interval.ms` is set to
30 seconds, the same as the default `commit.interval.ms`. Users who have
configured a different `commit.interval.ms` may need to review and
change `repartition.purge.interval.ms`.

Unlike `commit.interval.ms`, we don't dynamically change the default for
`repartition.purge.interval.ms` when EOS processing is enabled, as it's
important not to flood brokers with the record deletions, and we want a
sensible default.

nicktelford commented:
@mjsax

mjsax commented:
Thanks for the PR. Can you do a KIP for this? Happy to review the PR after the KIP is approved.

nicktelford commented:
KIP available at: https://cwiki.apache.org/confluence/display/KAFKA/KIP-811%3A+Add+separate+min.repartition.purge.interval.ms+to+Kafka+Streams

The commit and PR description have been updated to the latest specification from the KIP, based on the dev list discussion.

nicktelford commented:
@mjsax The KIP has now been adopted and I've updated this PR to reflect the latest version of the KIP. Should now be ready for review.

mjsax reviewed (commented):
The fix itself LGTM. We should add a test to `StreamThreadTest` though -- it's already setup to use a mocked `TaskManager` and mocked `Time` so it should not be too hard to verify that `maybePurgeCommittedRecords` is only called after the purge interval passed.

cadonna commented:
@nicktelford and @mjsax Do you think we can get this PR merged by March 16th so that we can ship this feature with AK 3.2.0?

nicktelford commented:
@cadonna I believe the only missing piece is the test(s), which should be done well before March 16th.

nicktelford commented:
@mjsax I don't see a `StreamThreadTest`. Do you mean `GlobalStreamThreadTest`?

nicktelford committed:
Add tests for repartition purge interval

Two tests are added, along with a utility method to construct the mock
`TaskManager`:

  * A test that purges don't occur before the repartition interval, even
    after a commit.
  * A test that purges occur after the repartition interval, once
    records have been committed.

mjsax commented:
Merged to `trunk`. Thanks for the KIP, PR, and for adding test!

Very nice addition!
